src/types/Utf8Utf16Convert.cpp

```diff
@@ -208,161 +208,30 @@ void u16state::reset() noexcept
-                                    ((gsl::narrow_cast<uint8_t>(*(it8 + 1)) ^ 0x00000080u) << 6u) |
-                                    (gsl::narrow_cast<uint8_t>(*(it8 + 2)) ^ 0x00000080u);
-                    }
-                }
-
-                it8 += cnt;
-                if (cnt < 3u)
-                {
-                    hRes = S_FALSE;
-                }
-            }
-            // valid four bytes
-            // - F0     | 90..BF | 80..BF | 80..BF
-            // - F1..F3 | 80..BF | 80..BF | 80..BF
-            // - F4     | 80..8F | 80..BF | 80..BF
-            else if (gsl::narrow_cast<uint8_t>(*it8) >= 0xF0u && gsl::narrow_cast<uint8_t>(*it8) <= 0xF4u)
-            {
-                size_t cnt{ 1u };
-                if ((it8 + 1) < end8 &&
-                    (( // F0     | *90*..BF
-                         gsl::narrow_cast<uint8_t>(*it8) == 0xF0u &&
-                         gsl::narrow_cast<uint8_t>(*(it8 + 1)) >= 0x90u && gsl::narrow_cast<uint8_t>(*(it8 + 1)) <= contEnd) ||
-                     ( // F1..F3 | 80..BF
-                         gsl::narrow_cast<uint8_t>(*it8) >= 0xF1u && gsl::narrow_cast<uint8_t>(*it8) <= 0xF3u &&
-                         gsl::narrow_cast<uint8_t>(*(it8 + 1)) >= contBegin && gsl::narrow_cast<uint8_t>(*(it8 + 1)) <= contEnd) ||
-                     ( // F4     | 80..*8F*
-                         gsl::narrow_cast<uint8_t>(*it8) == 0xF4u &&
-                         gsl::narrow_cast<uint8_t>(*(it8 + 1)) >= contBegin && gsl::narrow_cast<uint8_t>(*(it8 + 1)) <= 0x8Fu)))
-                {
-                    ++cnt;
-                    if ((it8 + 2) < end8 && gsl::narrow_cast<uint8_t>(*(it8 + 2)) >= contBegin && gsl::narrow_cast<uint8_t>(*(it8 + 2)) <= contEnd)
-                    {
-                        ++cnt;
-                        if ((it8 + 3) < end8 && gsl::narrow_cast<uint8_t>(*(it8 + 3)) >= contBegin && gsl::narrow_cast<uint8_t>(*(it8 + 3)) <= contEnd)
-                        {
-                            ++cnt;
-                            codePoint = ((gsl::narrow_cast<uint8_t>(*it8) ^ 0x000000F0u) << 18u) |
-                                        ((gsl::narrow_cast<uint8_t>(*(it8 + 1)) ^ 0x00000080u) << 12u) |
-                                        ((gsl::narrow_cast<uint8_t>(*(it8 + 2)) ^ 0x00000080u) << 6u) |
-                                        (gsl::narrow_cast<uint8_t>(*(it8 + 3)) ^ 0x00000080u);
-                        }
-                    }
-                }
+        int lengthRequired{};
+        RETURN_HR_IF(E_ABORT, FAILED(SizeTToInt(in.length(), &lengthRequired)));
 
-                it8 += cnt;
-                if (cnt < 4u)
-                {
-                    hRes = S_FALSE;
-                }
-            }
-            else
-            {
-                hRes = S_FALSE;
-                ++it8;
-            }
+        out.resize(in.length()); // avoid to call MultiByteToWideChar twice only to get the required size

```

Comment:
Did you measure how much worse the performance is with the double call? I'm curious. I've always done the double-call as a safety thing in case they know something I don't.

Reply:
I recalled to perform these tests but didn't find the code anymore.
However, that are the results.  
..._worstCase uses the worst case length (current implementation)  
..._determineLength calls the API function twice, to get the required length and then perform the conversion
```~~~
test "CompNaturalLang_Chunks - en.txt"
 u8u16_worstCase length       96000000 elapsed 3.5621
 u16u8_worstCase length       96000000 elapsed 3.44074
 u8u16_determineLength length 96000000 elapsed 4.63183
 u16u8_determineLength length 96000000 elapsed 4.4248

~~~
test "CompNaturalLang_Chunks - fr.txt"
 u8u16_worstCase length       96000000 elapsed 3.78793
 u16u8_worstCase length       100500000 elapsed 3.56667
 u8u16_determineLength length 96000000 elapsed 5.04372
 u16u8_determineLength length 100500000 elapsed 4.70454

~~~
test "CompNaturalLang_Chunks - ru.txt"
 u8u16_worstCase length       96000000 elapsed 5.06262
 u16u8_worstCase length       167400000 elapsed 4.4118
 u8u16_determineLength length 96000000 elapsed 7.37263
 u16u8_determineLength length 167400000 elapsed 6.00668

~~~
test "CompNaturalLang_Chunks - zh.txt"
 u8u16_worstCase length       96000000 elapsed 5.52145
 u16u8_worstCase length       247800000 elapsed 4.66172
 u8u16_determineLength length 96000000 elapsed 9.05221
 u16u8_determineLength length 247800000 elapsed 6.22064```